solidity/contracts/bridge/Bridge.sol

```diff
@@ -794,10 +934,571 @@ contract Bridge is Ownable {
+        require(mainUtxoHash != bytes32(0), "No main UTXO for given wallet");
+
+        // Assert that passed main UTXO parameter is the same as in storage and
+        // can be used for further processing.
+        require(
+            keccak256(
+                abi.encodePacked(
+                    mainUtxo.txHash,
+                    mainUtxo.txOutputIndex,
+                    mainUtxo.txOutputValue
+                )
+            ) == mainUtxoHash,
+            "Invalid main UTXO data"
+        );
+
+        // Assert that the single redemption transaction input actually
+        // refers to the wallet's main UTXO.
+        (bytes32 redemptionTxInputTxHash, uint32 redemptionTxInputTxIndex) =
+            processRedemptionTxInput(redemptionTxInputVector);
+        require(
+            mainUtxo.txHash == redemptionTxInputTxHash &&
+                mainUtxo.txOutputIndex == redemptionTxInputTxIndex,
+            "Redemption transaction input must point to the wallet's main UTXO"
+        );
+    }
+
+    /// @notice Processes the Bitcoin redemption transaction input vector. It
+    ///         extracts the single input then the transaction hash and output
+    ///         index from its outpoint.
+    /// @param redemptionTxInputVector Bitcoin redemption transaction input
+    ///        vector. This function assumes vector's structure is valid so it
+    ///        must be validated using e.g. `BTCUtils.validateVin` function
+    ///        before it is passed here
+    /// @return inputTxHash 32-byte hash of the Bitcoin transaction which is
+    ///         pointed in the input's outpoint.
+    /// @return inputTxIndex 4-byte index of the Bitcoin transaction output
+    ///         which is pointed in the input's outpoint.
+    function processRedemptionTxInput(bytes memory redemptionTxInputVector)
+        internal
+        pure
+        returns (bytes32 inputTxHash, uint32 inputTxIndex)
+    {
+        // To determine the total number of redemption transaction inputs,
+        // we need to parse the compactSize uint (VarInt) the input vector is
+        // prepended by. That compactSize uint encodes the number of vector
+        // elements using the format presented in:
+        // https://developer.bitcoin.org/reference/transactions.html#compactsize-unsigned-integers
+        // We don't need asserting the compactSize uint is parseable since it
+        // was already checked during `validateVin` validation.
+        // See `BitcoinTx.inputVector` docs for more details.
+        (, uint256 inputsCount) = redemptionTxInputVector.parseVarInt();
+        require(
+            inputsCount == 1,
+            "Redemption transaction must have a single input"
+        );
+
+        bytes memory input = redemptionTxInputVector.extractInputAtIndex(0);
+
+        inputTxHash = input.extractInputTxIdLE();
+
+        inputTxIndex = BTCUtils.reverseUint32(uint32(input.extractTxIndexLE()));
+
+        return (inputTxHash, inputTxIndex);
```

Comment:
```suggestion
        return (outpointTxHash, outpointIndex);
```
? 

I was confused about `inputTxIndex`. It is not an index of a transaction but an index of the previous output UTXO (our current input UTXO) in the last BTC transaction that produced the UTXO. It is... complicated, and I think it's easier if we just say "outpoint" given that it's explained here:

https://github.com/keep-network/tbtc-v2/blob/1e90ffa13d21022c37d1a2f5671906be77d1f247/solidity/contracts/bridge/BitcoinTx.sol#L34-L38

Reply:
And then, we should consider renaming `redemptionTxInputTxIndex` in `validateRedemptionTxInput` to just `redemptionTxInputIndex`

Reply:
There is a problem here because `outpointIndex` is also confusing. That suggests there are multiple outpoints in the input. The only thing that really makes sense is `inputTxOutputIndex` but it is intricate as well.

Regarding `redemptionTxInputTxIndex`, we cannot make it `redemptionTxInputIndex` since it means "the index of the redemption tx input" but it should mean  "the index of the transaction output pointed by the redemption tx input".

Reply:
See a8a55118